FWCore/Integration/test/ThinningTestAnalyzer.cc

```diff
@@ -174,24 +211,63 @@ namespace edmtest {
 
     std::vector<int>::const_iterator expectedValue = expectedValues_.begin();
     for (auto const& track : *trackCollection) {
-      if (*expectedValue == -1) {
+      if (not refToParentIsAvailable_ or *expectedValue == -1) {
         if (track.ref1.isAvailable()) {
-          throw cms::Exception("TestFailure") << "ref1 is available when it should not be";
+          throw cms::Exception("TestFailure") << "ref1 is available when it should not be, refers to "
+                                              << track.ref1.id() << " key " << track.ref1.key();
         }
       } else {
         if (!track.ref1.isAvailable()) {
           throw cms::Exception("TestFailure") << "ref1 is not available when it should be";
         }
         // Check twice to test some possible caching problems.
-        if (track.ref1->a != *expectedValue + eventOffset) {
-          throw cms::Exception("TestFailure") << "Unexpected values from ref1";
+        const int expected = (*expectedValue + eventOffset) * refSlimmedValueFactor_;
+        if (track.ref1->a != expected) {
+          throw cms::Exception("TestFailure")
+              << "Unexpected values from ref1, got " << track.ref1->a << " expected " << expected;
+        }
+        if (track.ref1->a != expected) {
+          throw cms::Exception("TestFailure")
+              << "Unexpected values from ref1 (2nd try), got " << track.ref1->a << " expected " << expected;
+          ;
         }
-        if (track.ref1->a != *expectedValue + eventOffset) {
-          throw cms::Exception("TestFailure") << "Unexpected values from ref1 (2nd try)";
+      }
+
+      if (not thinnedWasDropped_) {
+        auto refToThinned = edm::thinnedRefFrom(track.ref1, thinnedRefProd, event.productGetter());
```

Comment:
Given how big this is, maybe breaking each section into different member functions would help.

Reply:
I agree, but will leave to a subsequent PR.